### ISSUE TITLE
How to reproduce Figure 1?

### ISSUE THREAD

**Issue by Cranial-XIX** [berlino/seq_icl#10]
Hi authors,

Thanks for the great work. I am very interested in your work and would like to give the RegBench and those sequence models a try. I wonder if you could further elaborate on how to reproduce Figure 1.

Specifically:

- How to run a specific model under a specific number of training examples? What evaluation metric should the TV distance be?
- How to plot the plots in Figure 1 using the saved checkpoints/statistics?
- How to modify those models and delve into their design?
- How to modify the task?

Really thank you very much in advance :)

**Comment by ekinakyurek** [comment]
Hi @Cranial-XIX,

I am sorry for the late reply. I was traveling.

- The current training script already logs `2*TVD` to wandb as `final_test/model_dfa_diff`
- I attached the command to run best configuration for each model and each dataset size in below. 
- Please see `src/models/sequence/simple_lm.py` for main model script, `src/models/sequence/rnn/retentionv2.py` for example model. For each new model you also need to create configuration as in `configs/experiment/dfa/retnet.yaml`, and you need to add it to registry `src/utils/registry.py`
- See `src/dataloaders/dfa.py`, `configs/pipeline/dfa.yaml`, `configs/experiment/dfa/base.yaml` and the registry ``src/utils/registry.py` for adding a new task.


```
python train.py experiment=dfa/h3 dataset.num_test_examples=1000 dataset.num_examples=40000 model.n_layer=8 model.d_model=128 optimizer.lr=0.00025 optimizer.weight_decay=0.1
python train.py experiment=dfa/gla dataset.num_test_examples=1000 dataset.num_examples=40000 model.n_layer=8 model.d_model=128 optimizer.lr=0.00025 optimizer.weight_decay=0.1 model.attn_cfg.n_heads=2
python train.py experiment=dfa/linear_transformer dataset.num_test_examples=1000 dataset.num_examples=40000 model.n_layer=12 model.d_model=256 optimizer.lr=0.00025 optimizer.weight_decay=0.1 model.attn_cfg.num_heads=2 model.attn_layer_idx='[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]'
python train.py experiment=dfa/hyena dataset.num_test_examples=1000 dataset.num_examples=40000 model.n_layer=12 model.d_model=128 optimizer.lr=0.00025 optimizer.weight_decay=0.1
python train.py experiment=dfa/rwkv dataset.num_test_examples=1000 dataset.num_examples=40000 model.n_layer=4 model.d_model=128 optimizer.lr=0.00025 optimizer.weight_decay=0.1
python train.py experiment=dfa/retnet dataset.num_test_examples=1000 dataset.num_examples=40000 model.n_layer=8 model.d_model=256 optimizer.lr=0.00025 optimizer.weight_decay=0.1 model.attn_cfg.n_heads=2
python train.py experiment=dfa/transformer dataset.num_test_examples=1000 dataset.num_examples=40000 model.n_layer=12 model.d_model=128 optimizer.lr=0.00025 optimizer.weight_decay=0.1 model.attn_cfg.num_heads=2 model.attn_layer_idx='[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]'
python train.py experiment=dfa/s4d dataset.num_test_examples=1000 dataset.num_examples=40000 model.n_layer=12 model.d_model=128 optimizer.lr=0.00025 optimizer.weight_decay=0.1
python train.py experiment=dfa/mamba dataset.num_test_examples=1000 dataset.num_examples=40000 model.n_layer=8 model.d_model=512 optimizer.lr=0.00025 optimizer.weight_decay=0.1
python train.py experiment=dfa/lstm dataset.num_test_examples=1000 dataset.num_examples=40000 model.n_layer=12 model.d_model=256 optimizer.lr=0.00025 optimizer.weight_decay=0.1
python train.py experiment=dfa/hyena dataset.num_test_examples=1000 dataset.num_examples=20000 model.n_layer=12 model.d_model=128 optimizer.lr=0.00025 optimizer.weight_decay=0.1
python train.py experiment=dfa/linear_transformer dataset.num_test_examples=1000 dataset.num_examples=20000 model.n_layer=12 model.d_model=256 optimizer.lr=0.00025 optimizer.weight_decay=0.1 model.attn_cfg.num_heads=2 model.attn_layer_idx='[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]'
python train.py experiment=dfa/mamba dataset.num_test_examples=1000 dataset.num_examples=20000 model.n_layer=4 model.d_model=512 optimizer.lr=0.00025 optimizer.weight_decay=0.1
python train.py experiment=dfa/retnet dataset.num_test_examples=1000 dataset.num_examples=20000 model.n_layer=8 model.d_model=256 optimizer.lr=0.00025 optimizer.weight_decay=0.1 model.attn_cfg.n_heads=2
python train.py experiment=dfa/gla dataset.num_test_examples=1000 dataset.num_examples=20000 model.n_layer=8 model.d_model=128 optimizer.lr=0.00025 optimizer.weight_decay=0.1 model.attn_cfg.n_heads=2
python train.py experiment=dfa/h3 dataset.num_test_examples=1000 dataset.num_examples=20000 model.n_layer=8 model.d_model=128 optimizer.lr=0.00025 optimizer.weight_decay=0.1
python train.py experiment=dfa/s4d dataset.num_test_examples=1000 dataset.num_examples=20000 model.n_layer=12 model.d_model=128 optimizer.lr=0.00025 optimizer.weight_decay=0.1
python train.py experiment=dfa/transformer dataset.num_test_examples=1000 dataset.num_examples=20000 model.n_layer=8 model.d_model=128 optimizer.lr=0.0001 optimizer.weight_decay=0.1 model.attn_cfg.num_heads=2 model.attn_layer_idx='[0, 1, 2, 3, 4, 5, 6, 7]'
python train.py experiment=dfa/lstm dataset.num_test_examples=1000 dataset.num_examples=20000 model.n_layer=2 model.d_model=256 optimizer.lr=0.00025 optimizer.weight_decay=0.01
python train.py experiment=dfa/rwkv dataset.num_test_examples=1000 dataset.num_examples=20000 model.n_layer=12 model.d_model=128 optimizer.lr=0.00025 optimizer.weight_decay=0.1
python train.py experiment=dfa/hyena dataset.num_test_examples=1000 dataset.num_examples=10000 model.n_layer=12 model.d_model=128 optimizer.lr=0.00025 optimizer.weight_decay=0.1
python train.py experiment=dfa/mamba dataset.num_test_examples=1000 dataset.num_examples=10000 model.n_layer=4 model.d_model=512 optimizer.lr=0.00025 optimizer.weight_decay=0.1
python train.py experiment=dfa/rwkv dataset.num_test_examples=1000 dataset.num_examples=10000 model.n_layer=4 model.d_model=128 optimizer.lr=0.00025 optimizer.weight_decay=0.01
python train.py experiment=dfa/s4d dataset.num_test_examples=1000 dataset.num_examples=10000 model.n_layer=4 model.d_model=128 optimizer.lr=0.0001 optimizer.weight_decay=0.1
python train.py experiment=dfa/h3 dataset.num_test_examples=1000 dataset.num_examples=10000 model.n_layer=2 model.d_model=256 optimizer.lr=0.00025 optimizer.weight_decay=0.1
python train.py experiment=dfa/gla dataset.num_test_examples=1000 dataset.num_examples=10000 model.n_layer=12 model.d_model=128 optimizer.lr=0.00025 optimizer.weight_decay=0.1 model.attn_cfg.n_heads=2
python train.py experiment=dfa/linear_transformer dataset.num_test_examples=1000 dataset.num_examples=10000 model.n_layer=12 model.d_model=256 optimizer.lr=0.00025 optimizer.weight_decay=0.1 model.attn_cfg.num_heads=2 model.attn_layer_idx='[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]'
python train.py experiment=dfa/lstm dataset.num_test_examples=1000 dataset.num_examples=10000 model.n_layer=2 model.d_model=256 optimizer.lr=0.00025 optimizer.weight_decay=0.01
python train.py experiment=dfa/transformer dataset.num_test_examples=1000 dataset.num_examples=10000 model.n_layer=8 model.d_model=128 optimizer.lr=0.0001 optimizer.weight_decay=0.1 model.attn_cfg.num_heads=2 model.attn_layer_idx='[0, 1, 2, 3, 4, 5, 6, 7]'
python train.py experiment=dfa/retnet dataset.num_test_examples=1000 dataset.num_examples=10000 model.n_layer=4 model.d_model=256 optimizer.lr=0.00025 optimizer.weight_decay=0.1 model.attn_cfg.n_heads=2
python train.py experiment=dfa/s4d dataset.num_test_examples=1000 dataset.num_examples=5000 model.n_layer=12 model.d_model=128 optimizer.lr=0.00025 optimizer.weight_decay=0.1
python train.py experiment=dfa/transformer dataset.num_test_examples=1000 dataset.num_examples=5000 model.n_layer=8 model.d_model=128 optimizer.lr=0.0001 optimizer.weight_decay=0.1 model.attn_cfg.num_heads=2 model.attn_layer_idx='[0, 1, 2, 3, 4, 5, 6, 7]'
python train.py experiment=dfa/retnet dataset.num_test_examples=1000 dataset.num_examples=5000 model.n_layer=4 model.d_model=256 optimizer.lr=0.00025 optimizer.weight_decay=0.1 model.attn_cfg.n_heads=2
python train.py experiment=dfa/mamba dataset.num_test_examples=1000 dataset.num_examples=5000 model.n_layer=8 model.d_model=512 optimizer.lr=0.00025 optimizer.weight_decay=0.1
python train.py experiment=dfa/lstm dataset.num_test_examples=1000 dataset.num_examples=5000 model.n_layer=2 model.d_model=256 optimizer.lr=0.00025 optimizer.weight_decay=0.01
python train.py experiment=dfa/rwkv dataset.num_test_examples=1000 dataset.num_examples=5000 model.n_layer=8 model.d_model=128 optimizer.lr=0.00025 optimizer.weight_decay=0.1
python train.py experiment=dfa/hyena dataset.num_test_examples=1000 dataset.num_examples=5000 model.n_layer=12 model.d_model=128 optimizer.lr=0.00025 optimizer.weight_decay=0.1
python train.py experiment=dfa/linear_transformer dataset.num_test_examples=1000 dataset.num_examples=5000 model.n_layer=8 model.d_model=128 optimizer.lr=0.00025 optimizer.weight_decay=0.1 model.attn_cfg.num_heads=1 model.attn_layer_idx='[0, 1, 2, 3, 4, 5, 6, 7]'
python train.py experiment=dfa/gla dataset.num_test_examples=1000 dataset.num_examples=5000 model.n_layer=2 model.d_model=128 optimizer.lr=0.00025 optimizer.weight_decay=0.1 model.attn_cfg.n_heads=2
python train.py experiment=dfa/h3 dataset.num_test_examples=1000 dataset.num_examples=5000 model.n_layer=2 model.d_model=256 optimizer.lr=0.00025 optimizer.weight_decay=0.1
python train.py experiment=dfa/h3 dataset.num_test_examples=1000 dataset.num_examples=2500 model.n_layer=2 model.d_model=128 optimizer.lr=0.00025 optimizer.weight_decay=0.1
python train.py experiment=dfa/gla dataset.num_test_examples=1000 dataset.num_examples=2500 model.n_layer=2 model.d_model=128 optimizer.lr=0.00025 optimizer.weight_decay=0.1 model.attn_cfg.n_heads=1
python train.py experiment=dfa/linear_transformer dataset.num_test_examples=1000 dataset.num_examples=2500 model.n_layer=4 model.d_model=128 optimizer.lr=0.00025 optimizer.weight_decay=0.01 model.attn_cfg.num_heads=1 model.attn_layer_idx='[0, 1, 2, 3]'
python train.py experiment=dfa/lstm dataset.num_test_examples=1000 dataset.num_examples=2500 model.n_layer=2 model.d_model=256 optimizer.lr=0.00025 optimizer.weight_decay=0.01
python train.py experiment=dfa/transformer dataset.num_test_examples=1000 dataset.num_examples=2500 model.n_layer=12 model.d_model=64 optimizer.lr=0.00025 optimizer.weight_decay=0.1 model.attn_cfg.num_heads=2 model.attn_layer_idx='[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]'
python train.py experiment=dfa/s4d dataset.num_test_examples=1000 dataset.num_examples=2500 model.n_layer=2 model.d_model=128 optimizer.lr=0.0001 optimizer.weight_decay=0.01
python train.py experiment=dfa/retnet dataset.num_test_examples=1000 dataset.num_examples=2500 model.n_layer=4 model.d_model=64 optimizer.lr=0.00025 optimizer.weight_decay=0.1 model.attn_cfg.n_heads=1
python train.py experiment=dfa/rwkv dataset.num_test_examples=1000 dataset.num_examples=2500 model.n_layer=4 model.d_model=256 optimizer.lr=0.00025 optimizer.weight_decay=0.1
python train.py experiment=dfa/mamba dataset.num_test_examples=1000 dataset.num_examples=2500 model.n_layer=4 model.d_model=512 optimizer.lr=0.00025 optimizer.weight_decay=0.1
python train.py experiment=dfa/hyena dataset.num_test_examples=1000 dataset.num_examples=2500 model.n_layer=12 model.d_model=128 optimizer.lr=0.00025 optimizer.weight_decay=0.1
python train.py experiment=dfa/gla dataset.num_test_examples=1000 dataset.num_examples=1000 model.n_layer=2 model.d_model=128 optimizer.lr=0.00025 optimizer.weight_decay=0.1 model.attn_cfg.n_heads=1
python train.py experiment=dfa/h3 dataset.num_test_examples=1000 dataset.num_examples=1000 model.n_layer=2 model.d_model=128 optimizer.lr=0.00025 optimizer.weight_decay=0.1
python train.py experiment=dfa/lstm dataset.num_test_examples=1000 dataset.num_examples=1000 model.n_layer=2 model.d_model=128 optimizer.lr=0.00025 optimizer.weight_decay=0.1
python train.py experiment=dfa/linear_transformer dataset.num_test_examples=1000 dataset.num_examples=1000 model.n_layer=4 model.d_model=64 optimizer.lr=0.00025 optimizer.weight_decay=0.1 model.attn_cfg.num_heads=1 model.attn_layer_idx='[0, 1, 2, 3]'
python train.py experiment=dfa/hyena dataset.num_test_examples=1000 dataset.num_examples=1000 model.n_layer=4 model.d_model=64 optimizer.lr=0.0001 optimizer.weight_decay=0.01
python train.py experiment=dfa/mamba dataset.num_test_examples=1000 dataset.num_examples=1000 model.n_layer=8 model.d_model=512 optimizer.lr=0.00025 optimizer.weight_decay=0.1
python train.py experiment=dfa/rwkv dataset.num_test_examples=1000 dataset.num_examples=1000 model.n_layer=4 model.d_model=256 optimizer.lr=0.0001 optimizer.weight_decay=0.01
python train.py experiment=dfa/retnet dataset.num_test_examples=1000 dataset.num_examples=1000 model.n_layer=8 model.d_model=256 optimizer.lr=0.00025 optimizer.weight_decay=0.1 model.attn_cfg.n_heads=2
python train.py experiment=dfa/s4d dataset.num_test_examples=1000 dataset.num_examples=1000 model.n_layer=2 model.d_model=128 optimizer.lr=0.0001 optimizer.weight_decay=0.1
python train.py experiment=dfa/transformer dataset.num_test_examples=1000 dataset.num_examples=1000 model.n_layer=2 model.d_model=128 optimizer.lr=0.0001 optimizer.weight_decay=0.01 model.attn_cfg.num_heads=1 model.attn_layer_idx='[0, 1]'
```

**Comment by Cranial-XIX** [comment]
Thanks a lot for this comprehensive response! Hope you had a great travel :)
